Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp

```diff
@@ -990,13 +988,138 @@ void WebAutomationSession::fragmentNavigatedForFrame(const WebFrameProxy& frame,
 {
     m_bidiProcessor->browsingContextDomainNotifier().fragmentNavigated(handleForWebFrameProxy(frame), navigationIDToProtocolString(navigationID), WallTime::now().secondsSinceEpoch().milliseconds(), frame.url().string());
 }
+
+void WebAutomationSession::contextCreatedForPage(const WebPageProxy& page)
+{
+    if (RefPtr mutablePage = WebProcessProxy::webPage(page.identifier())) {
+        mutablePage->getAllFrameTrees([this, protectedThis = Ref { *this }, pageID = page.identifier()](Vector<FrameTreeNodeData>&& trees) {
+            RefPtr page = WebProcessProxy::webPage(pageID);
+            if (!page)
+                return;
+
+            for (auto& tree : trees)
+                recursivelyEmitContextCreated(tree, std::nullopt);
+        });
+    }
+}
+
+std::pair<String, String> WebAutomationSession::getClientWindowAndUserContext(const WebPageProxy& page)
+{
+    String clientWindow = makeString(page.identifier().toUInt64());
+    String userContext = "default"_s;
+    return { clientWindow, userContext };
+}
+
+RefPtr<WebPageProxy> WebAutomationSession::getOpenerPage(const WebPageProxy& page)
+{
+    if (auto openerFrameID = page.openerFrameIdentifier()) {
+        if (RefPtr openerFrame = WebFrameProxy::webFrame(*openerFrameID))
+            return openerFrame->page();
+    }
+    return nullptr;
+}
+
+void WebAutomationSession::contextCreatedForFrame(const WebFrameProxy& frame)
+{
+    auto contextHandle = handleForWebFrameProxy(frame);
+    auto url = frame.url().string();
+    auto parentHandle = frame.parentFrame() ? handleForWebFrameProxy(*frame.parentFrame()) : "null"_s;
+
+    String clientWindow = "unknown-window"_s;
+    String userContext = "default"_s;
+
+    if (RefPtr page = frame.page()) {
+        auto [windowId, contextId] = getClientWindowAndUserContext(*page);
+        clientWindow = windowId;
+        userContext = contextId;
+    }
+
+    m_bidiProcessor->browsingContextDomainNotifier().contextCreated(contextHandle, url, "null"_s, parentHandle, JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>::create(), clientWindow, userContext);
```

Comment:
I'm not sure that we should be passing `null` here, which argument is that? Okay, I looked it up.. please use `let originalOpener` so it has a name.

I'm a little concerned that setting `"null"` won't actually set the field to `null` in the JSON payload, it will instead set a string. Have you looked into what it does? We might need to come up with a new solution for setting null fields in events, if it doesn't work already.

Reply:
FTR, this support can be added in a separate PR, keeping this one focused on the `contextCreated/Destroyed` events.